### ISSUE TITLE
[bug] conan install get wrong transient dependencies when using -o options

### ISSUE THREAD

**Issue by spektrof** [conan-io/conan#9132]
I would like to rebuild some necessary 3rd party libraries using different options, and create new packages with static or shared libraries. If I upload them to the remote repository and starting to consume them by specifying the options for the direct dependency of my project, then it does not take account those other options what I used at building. For example if I used the '_baz_' shared library to build the '_foo_' shared library, and then I am calling '_conan install foo/0.1@ -o foo:shared=True_', it gets the _static baz library_ instead of the _shared baz library_ despite of the conan remote what I could inspect stores the right package id for each transient dependencies.

I cannot use _*:shared=True_ because building a shared library does not mean that their transient dependencies will be shared libraries, some of them could be static libraries.

This makes the conflict resolution much harder because the only way to verify the packages is to check the _conaninfo.txt_ for each package after I downloaded them. It would be great if the _lockfile_ would reflect those options what used when creating the package. Also if I do not check this, then my application could crash with _error while loading shared libraries_.

I would assume that if I get a package using conan install, then every transient dependencies will be correct and can run any executable without any problem.

### Environment Details

  * Operating System+version: Ubuntu 18.04.5 LTS
  * Compiler+version: gcc 7.5.0
  * Conan version: 1.37.2
  * Python version: 3.6.9

### Steps to reproduce

I followed [this](https://repository.prace-ri.eu/git/help/user/packages/conan_repository/index.md) tutorial to setup an own remote (called as gitlab) with some example project.
My public projects on gitlab: [conan-foo](https://gitlab.com/spektrof/conan-foo) and [conan-app](https://gitlab.com/spektrof/conan-app) where conan-app depends on conan-foo

Created and uploaded the packages after cloned the two projects:
```
$ cd conan-foo
$ conan create . spektrof+conan-foo/beta -o *:shared=False
$ conan create . spektrof+conan-foo/beta -o *:shared=True
$ cd ../conan-app
$ conan create . spektrof+conan-app/beta -o *:shared=False
$ conan create . spektrof+conan-app/beta -o *:shared=True
$ conan upload '*' -r gitlab --all
```

Get the uploaded artifacts after cleared the cache:
`$ conan install App/0.1@spektrof+conan-app/beta -o App:shared=True`

### Logs

Error when executing the app because it downloaded the foo static library:
```
$ /var/tmp/conan-data/App/0.1/spektrof+conan-app/beta/package/f664b29838aaf2bb0fbb11ed8707fde8a7c2e558/bin/app
/var/tmp/conan-data/App/0.1/spektrof+conan-app/beta/package/f664b29838aaf2bb0fbb11ed8707fde8a7c2e558/bin/app: error while loading shared libraries: libfoo.so: cannot open shared object file: No such file or directory
```

I attached the two conaninfo.txt for those packages what have been downloaded by conan install.

I generated the lockfile by:
`$ conan lock create --reference App/0.1@spektrof+conan-app/beta -o App:shared=True`

which also gives wrong option list for App, because I didn't even build App with that configuration, however the _package_id_ is good.

[app_conaninfo.txt](https://github.com/conan-io/conan/files/6687820/app_conaninfo.txt)
[foo_conaninfo.txt](https://github.com/conan-io/conan/files/6687640/foo_conaninfo.txt)
[conan-lock.txt](https://github.com/conan-io/conan/files/6687646/conan-lock.txt)

**Comment by memsharded** [comment]
Hi @spektrof 

I am not sure if I fully understand the issue. I think you might be misunderstanding the ``shared`` option for the ``app`` package. The ``app`` package doesn't have the ``shared`` option as it is not a library, neither static or shared, it is just an executable. It doesn't make sense. It will have instead 2 different binaries, for different options of the dependencies. So something like:

- The ``foo`` package has the options ``shared``, you create it as you do:
```
$ cd conan-foo
$ conan create . foo/beta@ -o foo:shared=False
$ conan create . foo/beta@ -o foo:shared=True
```

- The ``app`` package doesnt' have the ``shared`` option. You can create 2 different binaries for it with:

```
$ cd conan-app
$ conan create . app/beta@ -o foo:shared=False
$ conan create . app/beta@ -o foo:shared=True
$ conan search app/beta@  # should show 2 different package_ids
```

Then for installing the package you use:

```
$ conan install app/beta@ -o foo:shared=True
```

Please let me know if this makes sense.

**Comment by spektrof** [comment]
Hi @memsharded 

You are right, but the issue still exists if the app would build an own library beside the executable. This was the minimal example to reproduce my problem, sorry, it was a bit confusing.

The point is if I want to get a direct dependency, then I don't want to list all of its transitive dependencies with the proper shared/static options because that should be defined by the recipe and the uploaded package. The artifactory stores the right (dep, package_id) pairs for the direct dependency but I get different packages when installing it.

Back to the example the conaninfo of the downloaded app says that it has a dependency: 
`Foo/0.1@spektrof+conan-foo/beta:27cb7fbffba1a5268eda91c4da54b0254d48c3b1` which would be good, but actually I got other Foo (the static one).

**Comment by spektrof** [comment]
I have created an other project, called [conan-bar](https://gitlab.com/spektrof/conan-bar) which is a dependency of conan-foo now.

So what I did after clone:
```
$ conan create . spektrof+conan-bar/beta -o *:shared=False
$ conan create . spektrof+conan-bar/beta -o *:shared=True
$ conan create . spektrof+conan-foo/beta -o *:shared=False
$ conan create . spektrof+conan-foo/beta -o *:shared=True
$ ld /var/tmp/conan-data/Foo/0.1/spektrof+conan-foo/beta/package/0233e9be87ba84421884958faa02217c10c8f6cf/lib/libfoo.so -e 0x0000000000000980
$ conan upload '*' -r gitlab --all
```

Then after cleared the cache:
```
$ conan install Foo/0.1@spektrof+conan-foo/beta -o Foo:shared=True #  there can be tons of transitive deps so I dont want to list them here, the package in the artifactory stores the right ids

$ ld /var/tmp/conan-data/Foo/0.1/spektrof+conan-foo/beta/package/0233e9be87ba84421884958faa02217c10c8f6cf/lib/libfoo.so -e 0x0000000000000980
ld: warning: libbar.so, needed by /var/tmp/conan-data/Foo/0.1/spektrof+conan-foo/beta/package/0233e9be87ba84421884958faa02217c10c8f6cf/lib/libfoo.so, not found (try using -rpath or -rpath-link)
/var/tmp/conan-data/Foo/0.1/spektrof+conan-foo/beta/package/0233e9be87ba84421884958faa02217c10c8f6cf/lib/libfoo.so: undefined reference to `get_bar()'

$ conan get Foo/0.1@spektrof+conan-foo/beta:0233e9be87ba84421884958faa02217c10c8f6cf | grep Bar
    Bar/0.1
    Bar/0.1@spektrof+conan-bar/beta:27cb7fbffba1a5268eda91c4da54b0254d48c3b1
    Bar:fPIC=False
    Bar:shared=True
$ conan search Bar/0.1@spektrof+conan-bar/beta
Existing packages for recipe Bar/0.1@spektrof+conan-bar/beta:

    Package_ID: d351525cc53ebe68279edf1978846402420066e7
        [options]
            fPIC: True
            shared: False
    ....
```

**Comment by memsharded** [comment]
Hi @spektrof 

The problem is still the same. If you want ``bar`` to be shared when installing it, you need to tell Conan:

```bash
conan install Foo/0.1@spektrof+conan-foo/beta -o Foo:shared=True -o Bar:shared=True
# or the same as you did when you created it
conan install Foo/0.1@spektrof+conan-foo/beta -o *:shared=True
```

> The point is if I want to get a direct dependency, then I don't want to list all of its transitive dependencies with the proper shared/static options because that should be defined by the recipe and the uploaded package. 

That is not how Conan works. First, the inputs are defined. If you want to use ``Bar:shared=True`` then you need to tell the system. Based on all the inputs, conan will compute the ``package_id`` of the binaries, only then, it will have access to the specific binary, with the specific ``conaninfo.txt`` details. But you could have 10 different ``package_ids`` for Foo in the cache, some were linked with different compilers and with different options of themselves and their dependencies. Every possible combination of install options will give  a different package_id.

There are 2 ways to define that:

- Define things in a profile file. This is the recommended way, and not command line arguments. For production, you should be using profiles.
- You can actually propose defaults in recipes, so in ``Foo``, you can put: ``default_options = {"Bar:shared": True}`` if that is the default you prefer. But in this case, the reverse will also happen, if you create a Foo version with ``Bar:shared=False``, then when you do ``conan install`` if you want that binary you will need to explicitly defin ``Bar:shared=False`` again.

What I don't fully get is why you aren't getting "Binary missing" errors, maybe you are ``--build=missing``, or you created the static library for ``Bar`` too?

**Comment by spektrof** [comment]
Hi @memsharded 

What if as a user I don't know which options were used at building? Also this means if some library linked with static libraries and other libraries linked with shared libraries, then I need to call `conan install` separately, because I cannot specify both `foo:shared=True` (which lets say required by `baz1`) and `foo:shared=False` (which required by `baz2`) in the same `conanfile.txt` or `profile` file. 

So I cannot have a `conanfile.txt`:
```
[required]
baz1/0.1
baz2/0.1
?
```
instead of that, I need:
```
baz1/conanfile.txt:
[required]
baz1/0.1
[options]
foo:shared=True

baz2/conanfile.txt:
[required]
baz2/0.1
[options]
foo:shared=False
```

Maybe this is the expected way to do that.

I am using `Bazel` build system with conan packages (which are installed by a Bazel rule) and would like to handle everything by using the minimal number of conan command without hard-coding those build options what I used to rebuild a 3rd-party package, so if the user ask for example shared boost libraries, then they get that version what I provided for them.
Currently I am checking each `conanfile.txt` after a `conan install` to check whether I got the right transitive dependencies or not which means that it takes more time to resolve the conflicts.

Its true that there can be 10 different `package_id` for lets say a shared library but if I didn't specify enough options, then the `conan` should throw an error. Its weird for me to get transitive dependency with that `package_id `which is not really required.

Using default options could help, but that means I should patch every recipe file from `conan-center-index`.

I got that error because I created static and shared libraries for `Bar`, however when installed `Foo` shared library (which depends on `Bar `shared library), but I got the `Bar `static libraries instead.
If I use `conan install Foo/0.1@spektrof+conan-foo/beta -o Foo:shared=True -o Bar:shared=True` what you mentioned, then it can find the `Baz` shared library and won't fail when calling `ld`.

**Comment by memsharded** [comment]
> What if as a user I don't know which options were used at building? 

Building and installing/consuming are separate and decoupled processes. When you install/consume you declare what you want, the configuration you need (the inputs), that include how you want your dependencies (shared, static), what compilers and version are you using, etc. If it happens that there are pre-compiled binaries for it, they will be used. If not, Conan will error saying "Missing binary for this configuration", or will build from sources if ``--build=missing``. The flow is not "I build this binary called B1, and I consume this binary B1". This can be clearly seen with different types of packages, for example, header-only libraries will have exactly 1 ``package_id`` for all different configurations of compilers, dependencies, etc, as it is always the same headers. You don't build it for "release" then consume that binary for release. The same ``package_id`` will be consumed by many different combinations of settings and options. A static library with ``shared=False`` will use the same ``package_id`` of the same header-only library, as well as other packages with ``shared=True``.

> Also this means if some library linked with static libraries and other libraries linked with shared libraries, then I need to call conan install separately, because I cannot specify both foo:shared=True (which lets say required by baz1) and foo:shared=False (which required by baz2) in the same conanfile.txt or profile file.

Depending on how you define your dependencies, this is a conflict. In the same way you cannot depend both on ``zlib/1.2.8`` and ``zlib/1.2.11`` in the same conanfile or dependency graph, because that is a version conflict, you cannot depend on both ``Foo`` as static (``Foo:shared=False``) and shared (``Foo:shared=True``) binaries of the same package, that will be a configuration conflict, and Conan will raise an error.

A different use case is if you want to do 2 different builds of your application, then yes, you can call ``conan install`` twice, like:

```bash
$ conan install ....   -o Foo:shared=False
$ conan install ....    -o Foo:shared=True
```
(this is same for other variants, lets say, you want to install both Debug/Release dependencies, or you want to install x86/x86_64 dependencies, you do 1 ``conan install`` per configuration of dependencies you want to install).

Conan has mechanisms to manage multiple configurations from the consumer side, depending on the generator (if you are using multi-config build systems, like VS or CMake, you could be using the ``MSBuildDeps`` or ``CMakeDeps`` generators, which has customization for user custom configs). Or you could run those commands simply in different folders (typical flow for single-config build systems) to have different generated files in different folders.

**Comment by spektrof** [comment]
> will build from sources if --build=missing

I dont want to let the users to build packages from source.

> you cannot depend on both Foo as static (Foo:shared=False) and shared (Foo:shared=True) binaries of the same package

So if I want to link static boost/1.70.0 libraries for production but link shared boost/1.70.0 libraries for unit tests (because using static boost libs at testing may cause wasting of disk space), then I cannot create a single conanfile to my project and need to separate the core from the tests? I handled the version conflicts which is okay, but different units (like productions, tests) of my project could depend on different kind of libraries for some purpose.

> When you install/consume you declare what you want

This is true but only for direct dependencies, usually you don't need to bother with the transitive ones. Just imagine if your project has 20+ direct dependency and 100+ transitive dependencies, why should I specify the options for the transitive ones if the `package_id` of the direct dependencies were determined correctly?
I got your point, but in my view when consuming a project you only need to use these configuration values in order to get the proper `package_id`, which is working, and once its done you could use it to get the proper transitive dependencies from the `conaninfo.txt`.

> Conan has mechanisms to manage multiple configurations from the consumer side, depending on the generator

Probably I need to check that, at the moment I generate the files in different folders based on the platform and compiler.

> you do 1 conan install per configuration of dependencies you want to install

This is what I am doing with platforms and compilers, but I do not want to do that with my dependencies because I am using exactly one configurations for each dependency (so one to create static lib and one other to create shared lib).

**Comment by memsharded** [comment]
> I dont want to let the users to build packages from source.

Then, use a ``source()`` method that points to sources that are hidden from users, and only you has credentials to it. Otherwise, users can do ``--build`` to force building from sources.

> (because using static boost libs at testing may cause wasting of disk space)

Why? It is still the same package in the cache, it will not use extra disk space or anything like that. I am not sure what you mean here.

> I got your point, but in my view when consuming a project you only need to use these configuration values in order to get the proper package_id, which is working, and once its done you could use it to get the proper transitive dependencies from the conaninfo.txt.

This is not true, in the general case the ``package_id`` of a package is affected by transitive dependencies, not only by direct ones, So you need the transitive dependencies first, in order to compute the ``package_id``. So you need first to compute the whole dependency graph, including the options (shared=True/False) of every package, before being able to compute the ``package_id``. Access to the ``conaninfo.txt`` can only happen after knowing the options of the transitive dependencies, it cannot be evaluated in the other way, it is physically impossible. So the configuration, including settings and options for the graph need to be an input, either from profiles, or can be defined in ``default_options = {"Foo:shared": True}``, but it has to be defined, it is impossible to obtain it from conaninfo.txt.

**Comment by spektrof** [comment]
> I am not sure what you mean here.

https://www.boost.org/doc/libs/1_70_0/libs/test/doc/html/boost_test/usage_variants.html

I would like to link boost both statically and dynamically in my project, depending on the target.
In my opinion using static and shared libraries in the same project should be a viable scenario. But as far as I know this is not possible unless I modify the `conanfile` of the 3rd party lib in the `conan-center-index`.

> So you need first to compute the whole dependency graph, including the options (shared=True/False) of every package, before being able to compute the package_id

Okay, I agree with this. Then that means I had a unique case where I had only one configurations for creating static and one for shared libs but in general I cannot determine the `package_id` by declaring only the direct dependencies. It would be nice to verify the dependencies though to prevent these cases but that seems a hacky way.

Probably I will patch the `conanfiles` to set the `default_options`.

**Comment by memsharded** [comment]
> I would like to link boost both statically and dynamically in my project, depending on the target.

In one project you might use different variants of the same package, static and shared. As I commented above, for example, you might use ``build_requires`` and the ``build_context`` to put a different variant of boost in the ``build`` context. This is completely possible with Conan.

What is not possible is to have the same package to depend at the same time on both boost-static and boost-shared. Because how the build system integration will find them? What conan does for example for CMake with ``CMakeDeps`` generator is to generate a boost-config.cmake, so the ``find_package(boost)`` will find it. But you can only find 1 of them, not two, that is impossible from the build system point of view. You can have different builds, conditionals requirements, in which you can use the static or the shared one, but you cannot depend on the 2 variants of boost at the same time. This is not a Conan limitation, this cannot be done with bare CMake either.

> In my opinion using static and shared libraries in the same project should be a viable scenario.

It might be possible. It might also be possible to use different versions of the same library, for example you can use boost/1.64 and boost/1.65 in the same application, like building shared library wrappers around them, linking them statically into the shared ones, making sure no header propagates the public headers of the shared libs, and no object is passed through the APIs. I have seen that in production a few times. But that should be the very last resort, and done only because of legacy/debt, not by design. In general the static/shared library should be the same. Using 2 different binary variants of the same library in the same dependency graph is calling for linking errors (symbol conflicts) or in the worst case, very weird and difficult to debug ODR violations.


Having your real use case, instead of foo-bar examples, could throw a bit more of light into your use case. We are doing a lot of work in the Conan 2.0 new graph model, that explicitly models relations like "shared library links static library, hiding its details to consumers", so it would be good to understand what you are trying to do with boost in this case.

**Comment by spektrof** [comment]
Hi @memsharded 

> you might use build_requires and the build_context to put a different variant of boost in the build context

I am not quite sure how do you mean this here.

> more of light into your use case

As I mentioned above I am using [Bazel](https://bazel.build/) build system. This is great because its sandboxed, has great caching mechanism, clean, scalable, can define multiple platform, toolchain, compilers and the project libraries/applications easily and so on.

In `Bazel` you need to have a `WORKSPACE` file which declares your dependencies. These dependencies can be mapped from your filesystem with [new_local_repository](https://docs.bazel.build/versions/main/be/workspace.html#new_local_repository) or can use [source dependencies](https://docs.bazel.build/versions/main/repo/git.html) or even you can define your own [repository rule](https://docs.bazel.build/versions/main/skylark/repository_rules.html). This means that you could have multiple versions of the same library in your project, however its not really recommended as you said, I don't want that too.

In my case I want to use own repository rules to _form a dependency on a conan package_.
For example having the following in the `WORKSPACE`:

`conan_dependency(name="boost", version="1.70.0")`

I don't know all of my transient dependencies, I just want to declare boost, and want from the `Conan` side to get all of the required transient dependencies to be able to use it. I provide only one configuration for static and one configuration for shared libraries, so this should be not a problem.

Because its easy to have dependency conflicts due to having different versions of the same library which comes in as transitive dependency, I made a conflict resolution before these dependencies would be fetched. During this process I called the minimal number of conan calls to get all of my transient dependencies without any conflict.

I would like to be able to declare which kind of boost I want in the WOKRSPACE file. Such as:

```
conan_dependency(name="boost", version="1.70.0", shared=True)   # to get only shared libs
OR
conan_dependency(name="boost", version="1.70.0", static=False)    # to get only static libs
OR
conan_dependency(name="boost", version="1.70.0", shared=True, static=True)    # to get both shared and static libs
```

Each kind of library could have different transient dependencies, so I need to make sure about everything is correct after I resolved the conflicts. The easiest would be if I could call simple `conan install boost/1.70.0@ --profile my_profile -o boost:shared=True` , but as you said this is impossible because you have to start the evaluation of the package ids from the leaf nodes of the dep graph, except if I patch the original `conanfile `to add `default_options`. However I would suggest to have some kind of verification about that the right package ids were determined after this process (this should go from the root of the graph) because otherwise you could get `ld: warning: libbar.so, needed` errors what I described before. But you could say that, that's the user's fault.

It would be also good to get both shared libraries and static libraries with one command but I know that its not possible with the current restrictions, like they obviously could use different configurations so they should go to two different packages. This is fine.

Then back to `Bazel`, during [analysis phase](https://docs.bazel.build/versions/main/glossary.html) it will check your targets from [BUILD files](https://docs.bazel.build/versions/main/build-ref.html), where you put your [library, or binary rules](https://docs.bazel.build/versions/main/be/c-cpp.html).
When using pre-built libraries, you can use `cc_import` where you can specify each kind of library and you can "let" the `Bazel` to decide what to use at linking. 

For example:
```
cc_binary(
  name = "app",
  srcs = ["app.cpp"],
  deps = ["@boost//:boost_thread"],
  linkstatic = True # to link boost statically
)

cc_test(
  name = "lib_test",
  srcs = [...],
  deps = ["@boost//:boost_unit_test"],
  linkstatic = False # to link boost dynamically
)

cc_library(
  name = "boost_unit_test",
  srcs = [...headers...],
  deps = [":boost_unit_test_import"],
)

cc_import(
  name = "boost_unit_test_import",
  static_library = "lib/boost_unit_test_framework.a",
  shared_library = "lib/boost_unit_test_framework.so"
)

cc_library(
  name = "boost_thread",
  srcs = [...headers...],
  deps = [":boost_thread_import"],
)

cc_import(
  name = "boost_thread_import",
  static_library = "lib/boost_thread.a",
  shared_library = "lib/boost_thread.so"
)
```

You can of course specify whether you want static or shared linkage or you can say that this particular library should always linked statically or dynamically.

Most of the projects are modularized, and some modules could use static boost, while others could use shared boost or even both. Or what I mentioned previously, what if I want to use static boost libs normally, except for the unit tests where I want to link the boost unit test framework dynamically.

The point is that I need to have lets say static and shared boost libraries, so what I need to do by bazel rules is, get the transient dependencies of each kind of libraries what I declared in the `WORKSPACE` file for each available platforms (I dont want to use multiple versions of the same library), what I can do by running `conan info`, then I need to download them so I want the minimal number of `conan install` command because asynchronously I cannot use `conan install` as the `conan cache` could be malformed if two thread (from two different conan install call) are trying to download the same package, I am not sure about this, this was my experience. So I moved the package downloading to the same place to fix this issue and I am using `general.parallel_download` config.

We may want to create conan packages for our project, but don't want to create different packages just because I cannot have both static and shared library.

An other issue happens with the generator, because I cannot get the `cpp_info` of two differently configured package - correct me if I am wrong - so I cannot generate the proper `BUILD` files for those packages where I want to use both static and shared libraries. I have an idea how to fix this by `Bazel`.

We are planning to make these `bazel rules` for `conan_dependency` with the `build file generator` as an open source project but currently its not available.

Hope it helped a bit.

**Comment by memsharded** [comment]
Quick question: Are you aware that we released a new Bazel integration in Conan: https://docs.conan.io/en/latest/reference/conanfile/tools/google.html? It might be worth checking it, and contributing back to it, so we mature and have a standard built-in Bazel integration.

**Comment by spektrof** [comment]
I haven't seen that yet. Thanks for bringing it in, we will check.

However if I understand correctly this way the project can be only built by running `conan build`, I cannot do `bazel build` because that way it wont generate the build files for each dependency (or I need at least one `conan build`). This is something what we don't want, like why should I wrap the `bazel build` if I could use it directly?

We imagined some kind of similar way to use conan deps like https://github.com/bazelbuild/rules_jvm_external.

But my issue would still exist.
How to depend on both shared and static boost?
How those static and shared dependencies will be named? I want consistent naming for both such as `@boost//...` .
How the dependency resolution is working there? Should I list all of the transient dependencies? (because that's something what we don't want)

On the other hand it worth to try that out and give some feedback. Thanks again.

**Comment by memsharded** [comment]
> However if I understand correctly this way the project can be only built by running conan build, I cannot do bazel build because that way it wont generate the build files for each dependency (or I need at least one conan build). This is something what we don't want, like why should I wrap the bazel build if I could use it directly?

No, the new generators are intended to work with ``conan install`` + native build system, like ``bazel build``. The ``conan build`` will only be a convenience, but not mandatory (it will basically call ``bazel build`` under the hood, when you create a package).

With the new generators and ``generate()`` method, the necessary files for the native build will be generated at ``conan install``. I recommend to use the ``generate()`` method to automate the generation of files (this is what the new generators like BazelDeps do).

> How to depend on both shared and static boost?

It is not possible and it will not be possible to depend on both shared and static variants of the same library **in the same context**. Because you cannot link both variants at the same time. If building 2 different variants of the consumer, that is automatically managed by Conan, each consumer will be its own variant (or own package_id, but in any case this doesn't seem to be your use case). 

But it is possible to depend on both shared and static for your use case: the approach is the above mentioned: **use build and host contexts**. For the CMakeDeps generator we already enabled someway to generate files for packages in the build context, and there are ways to define a different name for them (the build_context_suffix in https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmakedeps.html#build-context-suffix)  (so they don't collide with the host context). This is also doable in the ``BazelDeps`` generator (and in the ``generate()`` method can be implemented manually too), but is not there yet, so I would suggest the custom ``generate()`` method iterating over ``self.dependencies``, and doing what you want based on the build-host context. 

This might not be straightforward, this ``self.dependencies`` is new, not documented, it will be better documented this next release, I hope that helps a bit.

**Comment by spektrof** [comment]
> generators are intended to work with conan install

True.

I moved far from my original "problem" which was actually was answered and I gave some suggestions.

I have a project which depends on several conan packages, but not necessarily has an own conanfile (maybe later it will have), this means that I don't want to call any `conan command` in order to build my project by `Bazel`. I want to use only `Bazel` and want that the rules do the conan things under the hood for me for each platform & compiler and be able to use remote execution at those platforms without refetching the whole project in the sandbox, all of these possible configurations are defined by our own toolchains and platforms so it can be done automatically. In that case what you suggest, I would need to call conan install with the generator with each configuration to generate the `BUILD` files and then I would be able to use my bazel commands.

I could generate a conanfile by a bazel rule for a simple projects where the package contained both static and shared libraries but thats not the intended way to do. This made me confused I guess, sorry.

I will still check that `bazel integration in conan` and the `build` and `host context` sooner or later.

**Comment by memsharded** [comment]
https://github.com/conan-io/conan/pull/10812 implemented a new integration for Bazel, to be released in 1.47, I recommend having a look.

**Comment by franramirez688** [comment]
Hi @spektrof 

Let's refresh this outdated ticket. We released a refactored integration with Bazel (Bazel + BazelToolchain + BazelDeps) available for Conan >= 1.62 and Conan >= 2.0.14. Any feedback would be more than welcome 😁 .

Apart from that, let's close this ticket like it did not have further discussions. Don't hesitate to reopen it (or open a new one) if any other problems come up.
